api/filrewardsd/main.go

```diff
@@ -0,0 +1,171 @@
+
+	rootCmd.PersistentFlags().StringVar(
+		&config.File,
+		"config",
+		"",
+		"Config file (default ${HOME}/"+config.Dir+"/"+config.Name+".yml)")
+	rootCmd.PersistentFlags().BoolP(
+		"debug",
+		"d",
+		config.Flags["debug"].DefValue.(bool),
+		"Enable debug logging")
+	rootCmd.PersistentFlags().String(
+		"logFile",
+		config.Flags["logFile"].DefValue.(string),
+		"Write logs to file")
+
+	rootCmd.PersistentFlags().String(
+		"listenAddr",
+		config.Flags["listenAddr"].DefValue.(string),
+		"Filrewards API listen address")
+
+	rootCmd.PersistentFlags().String(
+		"mongoUri",
+		config.Flags["mongoUri"].DefValue.(string),
+		"MongoDB connection URI")
+	rootCmd.PersistentFlags().String(
+		"mongoDb",
+		config.Flags["mongoDb"].DefValue.(string),
+		"MongoDB database name")
+
+	rootCmd.PersistentFlags().String(
+		"analyticsAddr",
+		config.Flags["analyticsAddr"].DefValue.(string),
+		"Analytics API address")
+
+	rootCmd.PersistentFlags().Int64(
+		"baseNanoFilReward",
+		config.Flags["baseNanoFilReward"].DefValue.(int64),
+		"Base NanoFIL reward amount",
+	)
+
+	err := cmd.BindFlags(config.Viper, rootCmd, config.Flags)
+	cmd.ErrCheck(err)
+}
+
+func main() {
+	cmd.ErrCheck(rootCmd.Execute())
+}
+
+var rootCmd = &cobra.Command{
+	Use:   daemonName,
+	Short: "Filrewards daemon",
+	Long:  `Textile's filrewards daemon.`,
+	PersistentPreRun: func(c *cobra.Command, args []string) {
+		config.Viper.SetConfigType("yaml")
+		cmd.ExpandConfigVars(config.Viper, config.Flags)
+
+		if config.Viper.GetBool("debug") {
+			err := util.SetLogLevels(map[string]logging.LogLevel{
+				daemonName: logging.LevelDebug,
+			})
+			cmd.ErrCheck(err)
+		}
+	},
+	Run: func(c *cobra.Command, args []string) {
+		settings, err := json.MarshalIndent(config.Viper.AllSettings(), "", "  ")
+		cmd.ErrCheck(err)
+		log.Debugf("loaded config: %s", string(settings))
+
+		debug := config.Viper.GetBool("debug")
+		logFile := config.Viper.GetString("log_file")
+		listenAddr := cmd.AddrFromStr(config.Viper.GetString("listen_addr"))
+		mongoUri := config.Viper.GetString("mongo_uri")
+		mongoDb := config.Viper.GetString("mongo_db")
+		analyticsAddr := config.Viper.GetString("analytics_addr")
+		baseFilReward := config.Viper.GetInt64("base_nano_fil_reward")
+
+		if logFile != "" {
+			err = cmd.SetupDefaultLoggingConfig(logFile)
+			cmd.ErrCheck(err)
+		}
+
+		target, err := util.TCPAddrFromMultiAddr(listenAddr)
+		cmd.ErrCheck(err)
+
+		listener, err := net.Listen("tcp", target)
+		cmd.ErrCheck(err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		conf := service.Config{
+			Listener:          listener,
+			MongoUri:          mongoUri,
+			MongoDbName:       mongoDb,
+			AnalyticsAddr:     analyticsAddr,
+			BaseNanoFILReward: baseFilReward,
+			Debug:             debug,
+		}
+		api, err := service.New(ctx, conf)
+		cmd.ErrCheck(err)
+
+		fmt.Println("Welcome to Hub Filrewards!")
+
+		cmd.HandleInterrupt(func() {
+			api.Close()
+			cancel()
```

Comment:
Since the service has a `Close` method, you could create the context internally, then cancel it in `Close`. That way the caller doesn't have to close and cancel.

Reply:
Ok, but what about the context that needs to be passed to `New` on line 158 above?

Reply:
Well, you wouldn't pass a context there. Create the context _inside_ the service.

Reply:
Ah now I see what you're saying. Cool, will update.